# Data Management

In the last lab, we introduced the basic commands and syntax in Python such as list, string, conditional statement, and for loop. We also became familiar with indexing and slicing a sequence, the frequently used built-in functions, and the if-statement and for-loop for iterating over a sequence. 

In this lab, we begin with the introduction of the Pandas for data management. You will be manipulating a new type of data called *DataFrame* which is the spreadsheet in Python. The dataset for this lab comes from a resale housing sample in Beijing from 2012 to 2016. 

Specifically. we will cover the following contents in this lab:


<p><a id="table"> </a></p>
<h3 id="Table-of-Contents">Table of Contents<a class="anchor-link" href="#Table-of-Contents">¶</a></h3>

<ol>
<h3>1 <a href="#pythonpackages">Python Packages</a><br/></h3>
</ol>

<ol>
<h3>2 <a href="#pandasbasics">Pandas basics</a><br/></h3>
2.1 <a href="#seriesanddataFrame">Series and DataFrame</a><br/>
2.2 <a href="#importandexportdata">Import and export data</a><br/>  
2.3 <a href="#indexingandslicingofdataFrame">Indexing and slicing DataFrame</a><br/>
2.4 <a href="#concatingmultipledataFramesalongtherows">Concatenating multiple DataFrames along the rows</a><br/>
2.5 <a href="#mergingdataframesalongthecolumns">Merging DataFrames along the columns</a><br/>
2.6 <a href="#moreaboutdatamanipulation">More about DataFrame manipulation</a><br/>
2.7 <a href="#assignment">Assignment</a><br/>    
    
</ol>



<p><a id="pythonpackages"> </a></p>
<h2 id="1-Python-packages">1 <a href="#table">Python packages</a><a class="anchor-link" href="#1-python-packages">¶</a></h2>

**Built-in Packages**: 
- Python comes with many built-in packages (also referred to as *modules* or *libraries*) that offer ready-to-use solutions to common programming problems. 
- These packages have already been installed with ArcGIS Pro (though Conda), and can be accessed using the *import* keyword. For today's lab, Pandas (more on this package below) can be imported by typing `import pandas as pd` in your current script.

- Here we load this package by `import` and give it a name "pd" using `as`, so that we do not need to type "pandas" each time we use it.

In [1]:
# import pandas and named it "pd" in the current python script
import pandas as pd

C:\Users\jewel\miniconda3\lib\site-packages\numpy\_distributor_init.py:30: UserWarning: loaded more than 1 DLL from .libs:
C:\Users\jewel\miniconda3\lib\site-packages\numpy\.libs\libopenblas.FB5AE2TYXYH2IJRDKGDGQ3XBKLKTF43H.gfortran-win_amd64.dll
C:\Users\jewel\miniconda3\lib\site-packages\numpy\.libs\libopenblas64__v0.3.21-gcc_10_3_0.dll
  warnings.warn("loaded more than 1 DLL from .libs:"


**Other Packages**: 
- The Python third-party packages are the packages that offer ready-to-use functions to common programming problems, but are not installed with ArcGIS Pro. We need to *install* them before importing to current script. 
- installing a third-party package require you to clone the current environment in ArcGIS Pro, then create a new one and activate it. 
- **A Python virtual environment** is a tool for dependency management and project isolation. It allows packages to be installed locally in an isolated directory for a particular project, as opposed to being installed globally (i.e. as part of a system-wide Python).


<p><a id="pandasbasics"> </a></p>
<h2 id="2pandasbasics">2 <a href="#table">Pandas basics</a><a class="anchor-link" href="#2Pandasbasics">¶</a></h2>

- Pandas provides fast and easy functions for **reading** and **writing** different kinds of files (e.g., .xlsx, .csv, .pickle) and **managing data** in the form of DataFrame. 
- It is established based on another package called numpy&mdash;which supports the multi-dimensional arrays and the mathematical functions for scientific computing. 
- Pandas extends numpy and provides new data types&mdash; **Series** and **DataFrame**.

<p><a id="seriesanddataFrame"> </a></p>
<h3 id="seriesanddataFrame">2.1 <a href="#table">Series and DataFrame</a><a class="anchor-link" href="#seriesanddataFrame">¶</a></h3>

The primary objects (Data types) in pandas are:

- ***Series***, which represents a single column of data (one dimension). You can think of *Series* objects as fancier versions of *list* data type.
- ***DataFrame***, which is very similar to an Excel spreadsheet, which has rows and named columns. A DataFrame contains one or more Series and a name for each Series (multiple dimensions).


- Series can be created through list using **pd.Series(list)**.

<p><a id="importandexportdata"> </a></p>
<h3 id="importandexportdata">2.2 <a href="#table">Import and export data</a><a class="anchor-link" href="#importandexportdata">¶</a></h3>

- Pandas can import and export dataset in many data formats using the Pandas [I/O API tool](https://pandas.pydata.org/pandas-docs/stable/user_guide/io.html). 
- we can read (import) an Excel (.xlsx) file from a file folder and present it as a DataFrame in Python. Also we can write (export) the DataFrame to Excel file using Pandas. 
- Pandas can also read and write .json, .html, .pickle, and other formats. To check whether pandas can read your data format, click [here](https://pandas.pydata.org/pandas-docs/stable/user_guide/io.html). 

The reading and writing functions of Excel file are:

- `df = pd.read_excel('<file Path>/data.xlsx')`&mdash;import .xlsx file to DataFrame or Series object, and assign it to a variable called df. 
- `df.to_excel('<file path>/data.xlsx')`&mdash;export a DataFrame called df to excel file 


- How to write a file path (`'<file Path>/data.xlsx'`)?
    - An ***absolute file path*** describes how to access a given file or directory, starting from the root of the file system.
    - A ***relative file path*** is an incomplete file path that is joined to your current working directory to create an absolute file path. 
    

Let us take an example using the 2012 Beijing housing resale dataset. The dataset was collected from a private housing agency called Lianjia (http://bj.lianjia.com/) and then was cleaned and subsampled to 5,000 transaction records.    

- Absolute file path: Begins from the root of the file system:

In [2]:
# an absolute file path
df = pd.read_excel("C:/CRP5680_TA/Lecture_Feb9/housing_dataset/HouseBeijing2012.xlsx")

- To construct a relative file path, we need to know the current file path which is the file path that you created this jupyter notebook. If you do not memorize where you created it, you can use the **os** package and the **os.getcwd()** to obtain the working path of the current Jupyter notebook.     

In [3]:
# get the current file path using os.getcwd function
import os
os.getcwd()

'c:\\CRP5680_TA\\Lecture_Feb9'

- Now make up in the parenthese the rest of the absolute path&mdash;"housing_dataset/HouseBeijing2012.xlsx". The file path in the parenthese is called relative file path. Always remember: ***`the absolute path = the current path + the relative path`***.   

In [4]:
df_2012 = pd.read_excel("housing_dataset/HouseBeijing2012.xlsx")

After you successfully read the Excel file, the variable `df_2012` is a dataframe that stores the excel spreadsheet. 
- To check the top five rows, use **`df.head(5)`**
- To check the last five rows, use **`df.tail(5)`**
- To explore the shape (dimension) of the DataFrame, use **`df.shape()`**
- To explore the column names, use **`df.columns`**. Note that the `df.columns` returns an object rather than a list. To convert it to list, use: `list(df_2012.columns)`



In [5]:
# explore the dataframe:
df_2012.head(5)

HouseID  CommunityID  TotalPrice  TransYear  Bedroom  Livingroom  \
0  BJFT84326414          1544  1400010.56       2012        2           1   
1  BJCP84958845          2606  1800066.00       2012        3           2   
2  BJDX84905788          2264  1350038.34       2012        2           1   
3  BJFT00386624          3621  1800006.91       2012        2           1   
4  BJCY84713854          1127  1970019.58       2012        1           1   

   Bathroom    Size  FloorLevel  WinSouth  ...  \
0         1   69.68           1         0  ...   
1         2  129.00           4         0  ...   
2         1   88.83           3         1  ...   
3         1   98.69           3         0  ...   
4         1   53.66           4         0  ...   

                                      XIAOQUWEB  SchQuality  NumSubway1km  \
0  https://bj.lianjia.com/xiaoqu/1111027377493/           0             2   
1  https://bj.lianjia.com/xiaoqu/1111027380050/           0             0   
2  https://bj.lianjia.com/xiaoqu/1111027379274/           0             1   
3  https://bj.lianjia.com/xiaoqu/1111027382765/           0             1   
4  https://bj.lianjia.com/xiaoqu/1111027376538/           0             3   

   Dist2Subway  HospQuality    Dist2Hosp  NumHosp1km  NumBus200m     Dist2CBD  \
0    633.24007            9   1803.02071           0           1   9345.20091   
1   2284.09390            9   9154.80958           0           0  18298.50637   
2    667.21572            8  11158.05983           0           4  22480.82065   
3    939.29061            9   1698.79101           0          10  16309.85203   
4    476.28267            9    938.35742           2           0   8105.90581   

   Dist2Center  
0   7396.31505  
1  18632.22305  
2  20105.06770  
3  11427.48851  
4   7213.87518  

[5 rows x 30 columns]

<p><a id="indexingandslicingofdataFrame"> </a></p>
<h3 id="indexingandslicingofdataFrame">2.3 <a href="#table">Indexing and slicing DataFrame</a><a class="anchor-link" href="#indexingandslicingofdataFrame">¶</a></h3>

selecting a subset of a DataFrame using **Indexing** and **Slicing**. 
- *Indexing* means simply selecting particular row or column from a DataFrame. 
- *Slicing* means selecting multiple rows and columns

Three ways of selecting particular rows and columns of an DataFrame object&mdash; `df[]`, `df.loc[rows_label , columns_label]` and `df.iloc[row_position , column_position]`. 
- **label** and **position** 
- A *label* can be understood as one name in the column list or an index in the row index column (the column at far left). 
- A *position* denotes the corresponding position of column name or index in a sequence, starting from zero.   

**1. We can select elements based on `df[]`**: 
- `df["col_name_1"]` select the column named "col_name1", and return a Series.
- `df[["col_name_1", "col_name_5"]]`, select multiple columns together, and return a DataFrame.

In [6]:
# select elements based on columns: 
df_2012["HouseID"] # select the column "HouseID"

0       BJFT84326414 
1       BJCP84958845 
2       BJDX84905788 
3       BJFT00386624 
4       BJCY84713854 
            ...      
4995    BJTJ84718789 
4996    BJFT84287006 
4997    BJDC84781079 
4998    BJSJ85075781 
4999    BJCY84949599 
Name: HouseID, Length: 5000, dtype: object

**In class exercise**
- try type in `df[["col_name_1"]]`, how it differs from `df["col_name_1"]`?

**2. Selecting elements based on `df.loc[ rows_label , columns_label ]`**:


In [7]:
# select first 3 rows and column "CommunityID" on a dataframe
df_2012.loc[0:2,"CommunityID"]

0    1544
1    2606
2    2264
Name: CommunityID, dtype: int64

In [8]:
# select the first 3 rows, and the columns "col_name_m" and "col_name_n"
df_2012.loc[0:2,["HouseID","TotalPrice"]]

HouseID  TotalPrice
0  BJFT84326414   1400010.56
1  BJCP84958845   1800066.00
2  BJDX84905788   1350038.34

In [9]:
# select the rows with label (index) 3 and 5, and columns from "HouseID" to "TotalPrice"
df_2012.loc[[3,5], "HouseID":"TotalPrice"]

HouseID  CommunityID  TotalPrice
3  BJFT00386624          3621  1800006.91
5  BJFT85228189          2768  1280012.60

**In class exercise**

- Select the last two rows and the last three columns

**3. Selecting elements based on `df.iloc[ row_position , column_position ]`:**

In [10]:
# select first 3 rows and the second column
df_2012.iloc[0:3,1]

0    1544
1    2606
2    2264
Name: CommunityID, dtype: int64

In [11]:
# select first 3 rows and columns "HouseID", "TotalPrice"
df_2012.iloc[0:3, [list(df_2012.columns).index("HouseID"), list(df_2012.columns).index("TotalPrice")]]

HouseID  TotalPrice
0  BJFT84326414   1400010.56
1  BJCP84958845   1800066.00
2  BJDX84905788   1350038.34

In [12]:
# select all rows and the last 2 columns "Dist2CBD" and "Dist2Center"
df_2012.iloc[ : , -2:]

Dist2CBD  Dist2Center
0      9345.20091   7396.31505
1     18298.50637  18632.22305
2     22480.82065  20105.06770
3     16309.85203  11427.48851
4      8105.90581   7213.87518
...           ...          ...
4995  15849.26185  21363.35490
4996   6221.44255   6278.89885
4997   3779.07141   3505.28158
4998  21094.05154  15576.85068
4999   6672.24055   7184.44822

[5000 rows x 2 columns]

**In-class exercise**

- select the first three rows and the column "TransYear" and "Dist2Subway"



<p><a id="concatingmultipledataFramesalongtherows"> </a></p>
<h3 id="concatingmultipledataFramesalongtherows">2.4 <a href="#table">Concatenating multiple Dataframes along the rows</a><a class="anchor-link" href="#concatingmultipledataFramesalongtherows">¶</a></h3>

- Pandas provides various functions for easily combining together Series or DataFrame along the rows or columns axis. 
- Let us try to concatenate the 2012 and 2013 Beijing housing dataset along the rows using `pd.concat()`. 
- "concatenate along the rows" means joining DataFrame B to DataFrame A vertically using column names as concatenating/joining identifiers.

To concatenate two DataFrames along the rows, you are requried to check whether the joining identifiers (column names) are the same in both datasets. Read the two datasets you want to concatenate and print out their column names:  

In [13]:
df_2012 = pd.read_excel("housing_dataset/HouseBeijing2012.xlsx")
df_2013 = pd.read_excel("housing_dataset/HouseBeijing2013.xlsx")

list(df_2012.columns) == list(df_2013.columns)

False

After a closer examination to the datasets, we find the 2013 dataset is slightly different from the 2012 dataset due to the existence of ordinal number before each column name. We can simply assign the column names of 2012 DataFrame to the column names of 2013 DataFrame to make them identical. And then, apply `.concat()` to combine the two.  

In [14]:
col_2012 = df_2012.columns # get the column name of 2012 data
df_2013.columns = col_2012 # assign the column names of 2012 dataset to the column names of 2013 dataset

# Concate (Join) 2012 and 2013 dataset along the rows using pd.concat() 
df_comb_1213 = pd.concat([df_2012, df_2013], axis = 0) # axis = 0 means concating along rows. axis =1 means concating along columns
df_comb_1213

HouseID  CommunityID  TotalPrice  TransYear  Bedroom  Livingroom  \
0     BJFT84326414          1544  1400010.56       2012        2           1   
1     BJCP84958845          2606  1800066.00       2012        3           2   
2     BJDX84905788          2264  1350038.34       2012        2           1   
3     BJFT00386624          3621  1800006.91       2012        2           1   
4     BJCY84713854          1127  1970019.58       2012        1           1   
...             ...          ...         ...        ...      ...         ...   
4995  BJCY85295028          3506  3400028.52       2013        2           2   
4996  BJFT86470175          2974  2040030.00       2013        2           1   
4997  BJCY85539965          1680  1970005.96       2013        1           1   
4998  BJDX85632173          3293  3150106.60       2013        3           2   
4999  BJTJ86774022          1714  2120057.94       2013        2           1   

      Bathroom    Size  FloorLevel  WinSouth  ...  \
0            1   69.68           1         0  ...   
1            2  129.00           4         0  ...   
2            1   88.83           3         1  ...   
3            1   98.69           3         0  ...   
4            1   53.66           4         0  ...   
...        ...     ...         ...       ...  ...   
4995         2  111.48           2         0  ...   
4996         1   57.00           5         0  ...   
4997         1   43.94           4         1  ...   
4998         2  130.45           1         1  ...   
4999         1   94.38           3         1  ...   

                                         XIAOQUWEB  SchQuality  NumSubway1km  \
0     https://bj.lianjia.com/xiaoqu/1111027377493/           0             2   
1     https://bj.lianjia.com/xiaoqu/1111027380050/           0             0   
2     https://bj.lianjia.com/xiaoqu/1111027379274/           0             1   
3     https://bj.lianjia.com/xiaoqu/1111027382765/           0             1   
4     https://bj.lianjia.com/xiaoqu/1111027376538/           0             3   
...                                            ...         ...           ...   
4995  https://bj.lianjia.com/xiaoqu/1111027382459/           0             2   
4996  https://bj.lianjia.com/xiaoqu/1111027381044/           0             0   
4997  https://bj.lianjia.com/xiaoqu/1111027377814/           0             1   
4998  https://bj.lianjia.com/xiaoqu/1111027381865/           0             0   
4999  https://bj.lianjia.com/xiaoqu/1111027377893/           0             0   

      Dist2Subway  HospQuality    Dist2Hosp  NumHosp1km  NumBus200m  \
0       633.24007            9   1803.02071           0           1   
1      2284.09390            9   9154.80958           0           0   
2       667.21572            8  11158.05983           0           4   
3       939.29061            9   1698.79101           0          10   
4       476.28267            9    938.35742           2           0   
...           ...          ...          ...         ...         ...   
4995    718.61206            9   3072.19586           0           0   
4996   1096.16421            9   1841.03871           0           0   
4997    577.23111            7    292.66932           1           7   
4998   1183.40088            8   2576.24539           0           5   
4999   3936.33494            9   1667.86016           0           2   

         Dist2CBD  Dist2Center  
0      9345.20091   7396.31505  
1     18298.50637  18632.22305  
2     22480.82065  20105.06770  
3     16309.85203  11427.48851  
4      8105.90581   7213.87518  
...           ...          ...  
4995   7698.13101   8724.78408  
4996   7833.44901   4972.72498  
4997   5901.86525   6733.63486  
4998  11878.33320  12565.01538  
4999  18100.89223  23595.76940  

[10000 rows x 30 columns]

The combined dataset looks good! It has 10,000 rows by 30 columns. But each row still has the original index that ranges from 0 to 4999. Let us reindex the rows using the function `df.reset_index( drop, inplace )`. 

The arguments `drop = True` means dropping the original index column.

The argument `inplace = True` means that the dataframe is modified (reindexed) directly so that we do not need to assign a new variable to store the returned value. If `inplace = False` or using the default, the dataframe will not change after reindexing, and we have to assign a new variable to save the change

In [15]:
df_comb_1213.reset_index(drop = True, inplace = True) 
df_comb_1213

HouseID  CommunityID  TotalPrice  TransYear  Bedroom  Livingroom  \
0     BJFT84326414          1544  1400010.56       2012        2           1   
1     BJCP84958845          2606  1800066.00       2012        3           2   
2     BJDX84905788          2264  1350038.34       2012        2           1   
3     BJFT00386624          3621  1800006.91       2012        2           1   
4     BJCY84713854          1127  1970019.58       2012        1           1   
...             ...          ...         ...        ...      ...         ...   
9995  BJCY85295028          3506  3400028.52       2013        2           2   
9996  BJFT86470175          2974  2040030.00       2013        2           1   
9997  BJCY85539965          1680  1970005.96       2013        1           1   
9998  BJDX85632173          3293  3150106.60       2013        3           2   
9999  BJTJ86774022          1714  2120057.94       2013        2           1   

      Bathroom    Size  FloorLevel  WinSouth  ...  \
0            1   69.68           1         0  ...   
1            2  129.00           4         0  ...   
2            1   88.83           3         1  ...   
3            1   98.69           3         0  ...   
4            1   53.66           4         0  ...   
...        ...     ...         ...       ...  ...   
9995         2  111.48           2         0  ...   
9996         1   57.00           5         0  ...   
9997         1   43.94           4         1  ...   
9998         2  130.45           1         1  ...   
9999         1   94.38           3         1  ...   

                                         XIAOQUWEB  SchQuality  NumSubway1km  \
0     https://bj.lianjia.com/xiaoqu/1111027377493/           0             2   
1     https://bj.lianjia.com/xiaoqu/1111027380050/           0             0   
2     https://bj.lianjia.com/xiaoqu/1111027379274/           0             1   
3     https://bj.lianjia.com/xiaoqu/1111027382765/           0             1   
4     https://bj.lianjia.com/xiaoqu/1111027376538/           0             3   
...                                            ...         ...           ...   
9995  https://bj.lianjia.com/xiaoqu/1111027382459/           0             2   
9996  https://bj.lianjia.com/xiaoqu/1111027381044/           0             0   
9997  https://bj.lianjia.com/xiaoqu/1111027377814/           0             1   
9998  https://bj.lianjia.com/xiaoqu/1111027381865/           0             0   
9999  https://bj.lianjia.com/xiaoqu/1111027377893/           0             0   

      Dist2Subway  HospQuality    Dist2Hosp  NumHosp1km  NumBus200m  \
0       633.24007            9   1803.02071           0           1   
1      2284.09390            9   9154.80958           0           0   
2       667.21572            8  11158.05983           0           4   
3       939.29061            9   1698.79101           0          10   
4       476.28267            9    938.35742           2           0   
...           ...          ...          ...         ...         ...   
9995    718.61206            9   3072.19586           0           0   
9996   1096.16421            9   1841.03871           0           0   
9997    577.23111            7    292.66932           1           7   
9998   1183.40088            8   2576.24539           0           5   
9999   3936.33494            9   1667.86016           0           2   

         Dist2CBD  Dist2Center  
0      9345.20091   7396.31505  
1     18298.50637  18632.22305  
2     22480.82065  20105.06770  
3     16309.85203  11427.48851  
4      8105.90581   7213.87518  
...           ...          ...  
9995   7698.13101   8724.78408  
9996   7833.44901   4972.72498  
9997   5901.86525   6733.63486  
9998  11878.33320  12565.01538  
9999  18100.89223  23595.76940  

[10000 rows x 30 columns]

**Let us combine the five housing dataset (2012 to 2016) together!**

- The following code uses `for loop`, `if statement` and `pd.concat()` to combine 5 Beijing housing datasets from 2012 to 2016, and export the combined DataFrame to an Excel file. 
- Pay attention to the `+ str(year) +` when writing the relative file path. `str(year)` means transforming the `year = 2012, ..., 2016` from an integer to a string such that `year = "2012",...,"2016"`. 
- The `+` signs at the both ends link two parts of the path. As such, each integral year can be converted to string and forms the relative file path, for example, `"housing_dataset/HouseBeijing2012.xlsx"`    

In [16]:
df_combine = pd.DataFrame() # create a blank dataframe

for year in list(range(2012, 2017)): 
    
    # construct the relative file path
    df_temp = pd.read_excel("housing_dataset/HouseBeijing" + str(year) + ".xlsx")
    
    if year == 2012:
        col_2012 = df_temp.columns
        df_combine = df_temp # assign df_2012 to df_combine in the first round of loop.
        
    else:     
        df_temp.columns = col_2012
        df_combine = pd.concat([df_combine, df_temp])

df_combine.reset_index(drop = True, inplace = True) 
df_combine.to_excel("housing_dataset/Beijing_combine.xlsx")
df_combine

HouseID  CommunityID  TotalPrice  TransYear  Bedroom  Livingroom  \
0      BJFT84326414          1544  1400010.56       2012        2           1   
1      BJCP84958845          2606  1800066.00       2012        3           2   
2      BJDX84905788          2264  1350038.34       2012        2           1   
3      BJFT00386624          3621  1800006.91       2012        2           1   
4      BJCY84713854          1127  1970019.58       2012        1           1   
...              ...          ...         ...        ...      ...         ...   
24995   101092000000         3341  2840059.18       2016        2           2   
24996  BJCY91554898           607  2250027.60       2016        1           1   
24997   101100000000         3619  3930039.20       2016        2           1   
24998   101101000000         1289  2550028.00       2016        3           1   
24999  BJHD91150416          2462  4250038.68       2016        3           1   

       Bathroom    Size  FloorLevel  WinSouth  ...  \
0             1   69.68           1         0  ...   
1             2  129.00           4         0  ...   
2             1   88.83           3         1  ...   
3             1   98.69           3         0  ...   
4             1   53.66           4         0  ...   
...         ...     ...         ...       ...  ...   
24995         1   94.42           5         1  ...   
24996         1   48.56           4         0  ...   
24997         1   63.40           5         1  ...   
24998         1   89.00           2         1  ...   
24999         1   70.23           2         0  ...   

                                          XIAOQUWEB  SchQuality  NumSubway1km  \
0      https://bj.lianjia.com/xiaoqu/1111027377493/           0             2   
1      https://bj.lianjia.com/xiaoqu/1111027380050/           0             0   
2      https://bj.lianjia.com/xiaoqu/1111027379274/           0             1   
3      https://bj.lianjia.com/xiaoqu/1111027382765/           0             1   
4      https://bj.lianjia.com/xiaoqu/1111027376538/           0             3   
...                                             ...         ...           ...   
24995  https://bj.lianjia.com/xiaoqu/1111027381983/           0             1   
24996  https://bj.lianjia.com/xiaoqu/1111027375189/           0             2   
24997  https://bj.lianjia.com/xiaoqu/1111027382757/           0             2   
24998  https://bj.lianjia.com/xiaoqu/1111027376948/           0             0   
24999  https://bj.lianjia.com/xiaoqu/1111027379686/           1             2   

       Dist2Subway  HospQuality    Dist2Hosp  NumHosp1km  NumBus200m  \
0        633.24007            9   1803.02071           0           1   
1       2284.09390            9   9154.80958           0           0   
2        667.21572            8  11158.05983           0           4   
3        939.29061            9   1698.79101           0          10   
4        476.28267            9    938.35742           2           0   
...            ...          ...          ...         ...         ...   
24995    868.51668            9   1600.57128           0           2   
24996    416.47153            9   1395.08991           0           0   
24997    218.23739            8    382.18439           1           6   
24998   1898.20842            8  11101.69181           0           0   
24999    190.21749            9   3274.97059           0          22   

          Dist2CBD  Dist2Center  
0       9345.20091   7396.31505  
1      18298.50637  18632.22305  
2      22480.82065  20105.06770  
3      16309.85203  11427.48851  
4       8105.90581   7213.87518  
...            ...          ...  
24995  22187.94046  20070.94098  
24996   7486.44958   6620.63431  
24997   4175.85237   7671.78186  
24998  40265.43022  37722.90810  
24999  13601.06560   9271.20384  

[25000 rows x 30 columns]

<p><a id="mergingdataframesalongthecolumns"> </a></p>
<h3 id="mergingdataframesalongthecolumns">2.5 <a href="#table">Merging DataFrames along the columns</a><a class="anchor-link" href="#mergingdataframesalongthecolumns">¶</a></h3>


Let us learn a new function `pd.merge()` that helps to merge DataFrames along the columns. `merge along the columns` means merging DataFrame B to DataFrame A horizontally. Note that the `pd.concat()` can also be used to merge along columns by changing the argument `axis = 1`; however, it is not as popular as `pd.merge()`. Also note, the function `pd.merge()` can **ONLY** be used to merge along the columns.

In the `df_combine`, you may notice the column (variable) "District" is coded by numbers, 1,2,3... without acknowledging us the specific district names. There is a file called `"DistrictName.xlsx"` in the Canvas Lab 8 folder stores the district name. We will merge the DataFrame with distrcit name to the current DataFrame.  

To merge the two DataFrames, we need to identify the merging key (identifier) which is the common column existing in both DataFrames. Remember that in the last section when joining along the columns, all the column names are the joining keys. In this merging case, the merging key in the current DataFrame (df_combine) is the column "District" and the merging key in the district DataFrame is the column "DistrictID".  


Check the merging key in the `df_combine`: column "District".

Read the `DistrictName.xlsx` and examine the merging keys.

In [17]:
# read the District name file. 
df_dist = pd.read_excel("housing_dataset/DistrictName.xlsx")
df_dist.head(5) # DistrictID as merge key

DistrictID  DistName_En DistName_Ch
0           1    Dongcheng          东城
1           2      Xicheng          西城
2           3     Chaoyang          朝阳
3           4      Fengtai          丰台
4           5  Shijingshan         石景山

In [18]:
# merge the combined dataframe with the district name
# left_on: the left merge key, right_on: the right merge key
df_combine = pd.merge(df_combine, df_dist, left_on = "District", right_on = "DistrictID")
df_combine

HouseID  CommunityID  TotalPrice  TransYear  Bedroom  Livingroom  \
0      BJFT84326414          1544  1400010.56       2012        2           1   
1      BJFT00386624          3621  1800006.91       2012        2           1   
2      BJFT85228189          2768  1280012.60       2012        1           1   
3      BJFT84585957          3607   757014.00       2012        1           0   
4      BJFT84434389          3171  1043222.04       2012        1           1   
...              ...          ...         ...        ...      ...         ...   
24995  BJMT91499815          3806  1440075.08       2016        2           1   
24996   101101000000          542  3400083.95       2016        2           1   
24997   101101000000         2998  3600116.52       2016        3           1   
24998   101100000000         1728  1900028.00       2016        1           1   
24999   101101000000         3819  3760013.36       2016        3           1   

       Bathroom    Size  FloorLevel  WinSouth  ...  Dist2Subway  HospQuality  \
0             1   69.68           1         0  ...    633.24007            9   
1             1   98.69           3         0  ...    939.29061            9   
2             1   46.10           3         1  ...   1648.56576            8   
3             1   26.94           2         0  ...    418.59917            9   
4             1   57.96           2         0  ...   2425.86815            9   
...         ...     ...         ...       ...  ...          ...          ...   
24995         1   76.13           5         1  ...   6184.24524            7   
24996         1   89.45           3         1  ...   5118.03636            7   
24997         1  117.26           2         0  ...   6532.98242            7   
24998         1   61.00           1         1  ...   5386.81625            7   
24999         1   89.68           2         1  ...   4160.80382            9   

        Dist2Hosp  NumHosp1km  NumBus200m     Dist2CBD  Dist2Center  \
0      1803.02071           0           1   9345.20091   7396.31505   
1      1698.79101           0          10  16309.85203  11427.48851   
2      1249.75505           0           1  16997.95563  11497.33703   
3      1747.82244           0           5   7084.13849   7200.33810   
4      4863.04910           0           2  14488.67850  10848.98986   
...           ...         ...         ...          ...          ...   
24995  1485.51072           0           0  30490.26155  24995.63395   
24996  2120.75266           0           0  29434.18998  23945.13966   
24997   522.45348           1          22  30821.21620  25352.70127   
24998  1684.72825           0           0  29693.26899  24217.50164   
24999  5736.85203           0           0  27241.95612  21723.35849   

       DistrictID  DistName_En  DistName_Ch  
0               2      Xicheng           西城  
1               2      Xicheng           西城  
2               2      Xicheng           西城  
3               2      Xicheng           西城  
4               2      Xicheng           西城  
...           ...          ...          ...  
24995          13       Pinggu           平谷  
24996          13       Pinggu           平谷  
24997          13       Pinggu           平谷  
24998          13       Pinggu           平谷  
24999          13       Pinggu           平谷  

[25000 rows x 33 columns]

<p><a id="moreaboutdatamanipulation"> </a></p>
<h3 id="moreaboutdatamanipulation">2.6 <a href="#table">More about DataFrame manipulation</a><a class="anchor-link" href="#moreaboutdatamanipulation">¶</a></h3>


Now, we want to further manipulate the dataset to obtain more useful informaiton, for example, what is the average housing price? how many housing samples are located within 1km of the subway stations? and which is the closest community to the CBD? You will learn how to do these in pandas in this section.  

**1. calculate unit housing price** 

You may notice that the DataFrame only has the column "TotalPrice" representing the total amount of money spent for a house, whereas in many applications, e.g., hedonic housing regression, the normalized price (unitprice) delivers richer information. Let us add a new column called "UnitPrice", obtained using the column "`TotalPrice`" divided by the column "`Size`" (*TotalPrice/Size*).

In [19]:
# by default the new column "UnitPrice" will be appended to the end of the list:
df_combine["UnitPrice"] = df_combine["TotalPrice"] / df_combine["Size"] 
df_combine

HouseID  CommunityID  TotalPrice  TransYear  Bedroom  Livingroom  \
0      BJFT84326414          1544  1400010.56       2012        2           1   
1      BJFT00386624          3621  1800006.91       2012        2           1   
2      BJFT85228189          2768  1280012.60       2012        1           1   
3      BJFT84585957          3607   757014.00       2012        1           0   
4      BJFT84434389          3171  1043222.04       2012        1           1   
...              ...          ...         ...        ...      ...         ...   
24995  BJMT91499815          3806  1440075.08       2016        2           1   
24996   101101000000          542  3400083.95       2016        2           1   
24997   101101000000         2998  3600116.52       2016        3           1   
24998   101100000000         1728  1900028.00       2016        1           1   
24999   101101000000         3819  3760013.36       2016        3           1   

       Bathroom    Size  FloorLevel  WinSouth  ...  HospQuality   Dist2Hosp  \
0             1   69.68           1         0  ...            9  1803.02071   
1             1   98.69           3         0  ...            9  1698.79101   
2             1   46.10           3         1  ...            8  1249.75505   
3             1   26.94           2         0  ...            9  1747.82244   
4             1   57.96           2         0  ...            9  4863.04910   
...         ...     ...         ...       ...  ...          ...         ...   
24995         1   76.13           5         1  ...            7  1485.51072   
24996         1   89.45           3         1  ...            7  2120.75266   
24997         1  117.26           2         0  ...            7   522.45348   
24998         1   61.00           1         1  ...            7  1684.72825   
24999         1   89.68           2         1  ...            9  5736.85203   

       NumHosp1km  NumBus200m     Dist2CBD  Dist2Center  DistrictID  \
0               0           1   9345.20091   7396.31505           2   
1               0          10  16309.85203  11427.48851           2   
2               0           1  16997.95563  11497.33703           2   
3               0           5   7084.13849   7200.33810           2   
4               0           2  14488.67850  10848.98986           2   
...           ...         ...          ...          ...         ...   
24995           0           0  30490.26155  24995.63395          13   
24996           0           0  29434.18998  23945.13966          13   
24997           1          22  30821.21620  25352.70127          13   
24998           0           0  29693.26899  24217.50164          13   
24999           0           0  27241.95612  21723.35849          13   

       DistName_En  DistName_Ch  UnitPrice  
0          Xicheng           西城    20092.0  
1          Xicheng           西城    18239.0  
2          Xicheng           西城    27766.0  
3          Xicheng           西城    28100.0  
4          Xicheng           西城    17999.0  
...            ...          ...        ...  
24995       Pinggu           平谷    18916.0  
24996       Pinggu           平谷    38011.0  
24997       Pinggu           平谷    30702.0  
24998       Pinggu           平谷    31148.0  
24999       Pinggu           平谷    41927.0  

[25000 rows x 34 columns]

**2. unique(): obtain the unique values**

The combined dataset covers 25,000 housing samples in total, but what if we are interested in knowing how many districts and communities it covers? We can use `pd.unique()` to obtain the unique values in a **single column** (i.e., a Series)

In [20]:
df_combine["DistName_En"].unique()

array(['Xicheng', 'Haidian', 'Fengtai', 'Mentougou', 'Fangshan',
       'Dongcheng', 'Huairou', 'Miyun', 'Shunyi', 'Changping', 'Chaoyang',
       'Shijingshan', 'Pinggu'], dtype=object)

In [21]:
print("The number of districts:", len(df_combine["DistName_En"].unique()))
print("The number of communities:", len(df_combine["CommunityID"].unique()))

The number of districts: 13
The number of communities: 3141


**3. Filtering DataFrame** 

Let us filter the DataFrame based on specified conditions. First, we want to select those housing samples located within 1500m of the subway stations. Those units are usually marked as "subway housing" for which buyers are willing to pay more. We can use the following structure to select the houses qualified for the condition.

  - `df_subway = df_combine[df_combine["Dist2Subway"] <= 1500]`:
      - step1, `df_combine["Dist2Subway"] <= 1500` return a series with values of ***False*** or ***True*** (boolean type); 
      - step2, it is wrapped by `df_combine[]` and can return a subset of the candidate rows
      - step3, assign the returned DataFrame to a new dataframe called `df_subway`

In [22]:
df_subway = df_combine[ df_combine["Dist2Subway"] <= 1500 ]
df_subway

HouseID  CommunityID  TotalPrice  TransYear  Bedroom  Livingroom  \
0      BJFT84326414          1544  1400010.56       2012        2           1   
1      BJFT00386624          3621  1800006.91       2012        2           1   
3      BJFT84585957          3607   757014.00       2012        1           0   
5      BJFT84439200          1295  1150036.80       2012        1           1   
8      BJFT84784085           929  3680059.50       2012        2           2   
...              ...          ...         ...        ...      ...         ...   
24908   101092000000         3892  2380038.00       2016        3           1   
24910  BJFS91284775          3810  1580077.70       2016        2           1   
24911   101100000000         2970  1350050.00       2016        2           1   
24912  BJFS91676189          3859  2980018.72       2016        2           2   
24913   101092000000         3741  3300080.00       2016        3           2   

       Bathroom    Size  FloorLevel  WinSouth  ...  HospQuality    Dist2Hosp  \
0             1   69.68           1         0  ...            9   1803.02071   
1             1   98.69           3         0  ...            9   1698.79101   
3             1   26.94           2         0  ...            9   1747.82244   
5             1   55.20           2         1  ...            9   2889.03867   
8             1   87.75           2         1  ...            8   1196.95857   
...         ...     ...         ...       ...  ...          ...          ...   
24908         1   89.00           3         1  ...            9  13631.02944   
24910         2   82.30           1         0  ...            8  17726.25702   
24911         1   67.00           1         1  ...            8  15062.18883   
24912         1   92.96           3         1  ...            9  13154.11751   
24913         1  112.00           1         1  ...            9  12661.51235   

       NumHosp1km  NumBus200m     Dist2CBD  Dist2Center  DistrictID  \
0               0           1   9345.20091   7396.31505           2   
1               0          10  16309.85203  11427.48851           2   
3               0           5   7084.13849   7200.33810           2   
5               0          12  16736.13924  12090.55677           2   
8               0           0  13705.83152   8421.64220           2   
...           ...         ...          ...          ...         ...   
24908           0           0  27837.00588  23340.24525           5   
24910           0           2  32529.80786  28283.72582           5   
24911           0           3  34402.65743  29930.17112           5   
24912           0           2  26516.34133  22376.04481           5   
24913           0           0  26415.12681  22109.49364           5   

       DistName_En  DistName_Ch  UnitPrice  
0          Xicheng           西城    20092.0  
1          Xicheng           西城    18239.0  
3          Xicheng           西城    28100.0  
5          Xicheng           西城    20834.0  
8          Xicheng           西城    41938.0  
...            ...          ...        ...  
24908  Shijingshan          石景山    26742.0  
24910  Shijingshan          石景山    19199.0  
24911  Shijingshan          石景山    20150.0  
24912  Shijingshan          石景山    32057.0  
24913  Shijingshan          石景山    29465.0  

[20279 rows x 34 columns]

**4. Creating categorical variables and using df.groupby() functions** 

Next let us consider a more comprehensive example. Let us say I am interesting in comparing the average housing prices for the houses that are located at different distances from the nearest subway station.

- To achieve this goal, we firstly create the categorical variable, denoting the subway accessibility. We can define a categorical variable 'Subwaylevel' that categorizes the continuous variable "Dist2Subway" (distance to subway) into 3 levels, 0-500m, 500-1500m, and beyond 1500m.  
- Then use a function `pd.groupby()` to group houses by the categorical variable, and summarize the average price by each group for comparison. 
- The `pd.groupby()` is similar to the pivotal table in Excel. It involves some combination of splitting the object, applying a function, and combining the results. In our example, we split the DataFrame in terms of the categorical variable "SubwayDum", specify a function to calculate the average value of each group, and regroup data to form a new table with average UnitPrice, Dist2Subway, and housing size. 

- Note that **the `pd.groupby()` does not return a DataFrame — you need to call sum(), mean(), etc, or apply() a function** to make the return a DataFrame. Here we use `pd.groupy().mean()` which returns a dataframe and we assign a new name for it: `df_subset`. Also note: the variable `Subwaylevel` is used as the index in the returned DataFrame.   

In [23]:
# step 1: create a categorical variable called SubwayDum
df_combine['Subwaylevel'] = 'Level 5'

df_combine.loc[(df_combine['Dist2Subway'] >= 0) & 
                  (df_combine['Dist2Subway'] <= 500), 'Subwaylevel'] = 'Level 1'

df_combine.loc[(df_combine['Dist2Subway'] >= 500) & 
                  (df_combine['Dist2Subway'] <= 1000), 'Subwaylevel'] = 'Level 2'

df_combine.loc[(df_combine['Dist2Subway'] >= 1000) & 
                  (df_combine['Dist2Subway'] <= 1500), 'Subwaylevel'] = 'Level 3'

df_combine.loc[(df_combine['Dist2Subway'] >= 1500) & 
                  (df_combine['Dist2Subway'] <= 2000), 'Subwaylevel'] = 'Level 4'

In [24]:
df_combine

HouseID  CommunityID  TotalPrice  TransYear  Bedroom  Livingroom  \
0      BJFT84326414          1544  1400010.56       2012        2           1   
1      BJFT00386624          3621  1800006.91       2012        2           1   
2      BJFT85228189          2768  1280012.60       2012        1           1   
3      BJFT84585957          3607   757014.00       2012        1           0   
4      BJFT84434389          3171  1043222.04       2012        1           1   
...              ...          ...         ...        ...      ...         ...   
24995  BJMT91499815          3806  1440075.08       2016        2           1   
24996   101101000000          542  3400083.95       2016        2           1   
24997   101101000000         2998  3600116.52       2016        3           1   
24998   101100000000         1728  1900028.00       2016        1           1   
24999   101101000000         3819  3760013.36       2016        3           1   

       Bathroom    Size  FloorLevel  WinSouth  ...   Dist2Hosp  NumHosp1km  \
0             1   69.68           1         0  ...  1803.02071           0   
1             1   98.69           3         0  ...  1698.79101           0   
2             1   46.10           3         1  ...  1249.75505           0   
3             1   26.94           2         0  ...  1747.82244           0   
4             1   57.96           2         0  ...  4863.04910           0   
...         ...     ...         ...       ...  ...         ...         ...   
24995         1   76.13           5         1  ...  1485.51072           0   
24996         1   89.45           3         1  ...  2120.75266           0   
24997         1  117.26           2         0  ...   522.45348           1   
24998         1   61.00           1         1  ...  1684.72825           0   
24999         1   89.68           2         1  ...  5736.85203           0   

       NumBus200m     Dist2CBD  Dist2Center  DistrictID  DistName_En  \
0               1   9345.20091   7396.31505           2      Xicheng   
1              10  16309.85203  11427.48851           2      Xicheng   
2               1  16997.95563  11497.33703           2      Xicheng   
3               5   7084.13849   7200.33810           2      Xicheng   
4               2  14488.67850  10848.98986           2      Xicheng   
...           ...          ...          ...         ...          ...   
24995           0  30490.26155  24995.63395          13       Pinggu   
24996           0  29434.18998  23945.13966          13       Pinggu   
24997          22  30821.21620  25352.70127          13       Pinggu   
24998           0  29693.26899  24217.50164          13       Pinggu   
24999           0  27241.95612  21723.35849          13       Pinggu   

       DistName_Ch  UnitPrice  Subwaylevel  
0               西城    20092.0      Level 2  
1               西城    18239.0      Level 2  
2               西城    27766.0      Level 4  
3               西城    28100.0      Level 1  
4               西城    17999.0      Level 5  
...            ...        ...          ...  
24995           平谷    18916.0      Level 5  
24996           平谷    38011.0      Level 5  
24997           平谷    30702.0      Level 5  
24998           平谷    31148.0      Level 5  
24999           平谷    41927.0      Level 5  

[25000 rows x 35 columns]

In [25]:
# step 2: group and calculate average price based on SubwayDum. 
df_subset = df_combine.groupby('Subwaylevel')[["UnitPrice","Dist2Subway", "Size"]].mean()
df_subset

UnitPrice  Dist2Subway       Size
Subwaylevel                                      
Level 1      39843.178209   351.931325  80.859418
Level 2      39383.787596   729.150232  78.656222
Level 3      34066.922906  1206.519367  81.868928
Level 4      29596.897008  1710.102008  81.568808
Level 5      24723.162938  3389.825217  88.982905

We can clearly observe that the average housing prices decrease drastically with the declining subway accessibility level, whereas the housing size displays an opposite trend. This finding makes sense because the locations of good subway access are mostly in the urban core areas with higher population density and housing price. 